File: frontend/src/shared/utils/validation.ts
import { z } from "zod";
import { ClothingItem, ClothingItemSchema, OpenAIClothingAnalysis, OpenAIClothingAnalysisSchema, ClothingTypeEnum, BasicMetadata, ColorAnalysis, OutfitSchema, UserProfileSchema, type Outfit, type UserProfile, type AppError, SuccessResponse, ErrorResponse } from '@shared/types';
import unidecode from 'unidecode';

// Maximum lengths for strings
const MAX_TYPE_LENGTH = 32;
const MAX_SUBTYPE_LENGTH = 64;

// Get the enum values as a const array
const CLOTHING_TYPE_VALUES = ClothingTypeEnum.options;

// Type for the clothing type mappings
type ClothingTypeMapping = {
  [K in z.infer<typeof ClothingTypeEnum>]: readonly string[];
};

// Canonical type mappings with synonyms
const TYPE_MAPPINGS: ClothingTypeMapping = {
  shirt: ['t-shirt', 'tee', 'tshirt', 'tee shirt', 'v neck', 'v-neck', 'vneck', 'tank top', 'polo'],
  dress_shirt: ['button down', 'button-up', 'oxford', 'formal shirt', 'dress shirt'],
  pants: ['jeans', 'trousers', 'slacks', 'chinos', 'khakis', 'leggings', 'joggers'],
  shorts: ['short pants', 'athletic shorts', 'basketball shorts', 'running shorts', 'cargo shorts'],
  skirt: ['mini skirt', 'maxi skirt', 'pleated', 'a-line'],
  dress: ['gown', 'frock', 'sundress', 'cocktail dress', 'maxi', 'mini'],
  jacket: ['denim', 'blazer', 'bomber', 'parka', 'windbreaker', 'coat', 'raincoat'],
  sweater: ['sweatshirt', 'hoodie', 'cardigan', 'pullover', 'jumper'],
  shoes: ['boots', 'heels', 'flats', 'pumps'],
  dress_shoes: ['oxfords', 'derby', 'wingtip', 'monk strap', 'formal shoes'],
  loafers: ['slip-ons', 'penny loafers', 'tassel loafers'],
  sneakers: ['trainers', 'athletic shoes', 'running shoes', 'kicks'],
  accessory: ['bag', 'purse', 'scarf', 'hat', 'jewelry', 'watch'],
  other: []
} as const;

// Helper function to normalize strings
export const normalizeString = (input: string | null | undefined): string => {
  if (!input) return '';
  
  // Convert to lowercase and normalize unicode
  let normalized = unidecode(input.toLowerCase().trim());
  
  // Remove special characters but keep spaces and hyphens
  normalized = normalized.replace(/[^a-z0-9\s-]/g, '');
  
  // Collapse multiple spaces and hyphens
  normalized = normalized.replace(/[\s-]+/g, ' ');
  
  return normalized;
};

// Helper function to find the closest match using Levenshtein distance
const findClosestMatch = (input: string, options: string[]): string | null => {
  if (!input || !options.length) return null;
  
  const normalizedInput = normalizeString(input);
  let bestMatch = null;
  let bestScore = 0.8; // Minimum similarity threshold
  
  for (const option of options) {
    const normalizedOption = normalizeString(option);
    const score = calculateSimilarity(normalizedInput, normalizedOption);
    
    if (score > bestScore) {
      bestScore = score;
      bestMatch = option;
    }
  }
  
  return bestMatch;
};

// Levenshtein distance calculation
const calculateSimilarity = (str1: string, str2: string): number => {
  const track = Array(str2.length + 1).fill(null).map(() =>
    Array(str1.length + 1).fill(null));
  
  for (let i = 0; i <= str1.length; i += 1) {
    track[0][i] = i;
  }
  for (let j = 0; j <= str2.length; j += 1) {
    track[j][0] = j;
  }

  for (let j = 1; j <= str2.length; j += 1) {
    for (let i = 1; i <= str1.length; i += 1) {
      const indicator = str1[i - 1] === str2[j - 1] ? 0 : 1;
      track[j][i] = Math.min(
        track[j][i - 1] + 1,
        track[j - 1][i] + 1,
        track[j - 1][i - 1] + indicator
      );
    }
  }

  const maxLength = Math.max(str1.length, str2.length);
  return 1 - (track[str2.length][str1.length] / maxLength);
};

// Normalize clothing type
export const normalizeClothingType = (type: string | null | undefined): z.infer<typeof ClothingTypeEnum> => {
  if (!type) return 'other';
  
  const normalized = normalizeString(type);
  if (!normalized) return 'other';
  
  // Check direct matches first
  if (CLOTHING_TYPE_VALUES.includes(normalized as z.infer<typeof ClothingTypeEnum>)) {
    return normalized as z.infer<typeof ClothingTypeEnum>;
  }
  
  // Check all synonyms
  for (const [canonicalType, synonyms] of Object.entries(TYPE_MAPPINGS)) {
    if (synonyms.includes(normalized)) {
      return canonicalType as z.infer<typeof ClothingTypeEnum>;
    }
  }
  
  // Try fuzzy matching
  const allOptions = [...CLOTHING_TYPE_VALUES, ...Object.values(TYPE_MAPPINGS).flat()];
  const match = findClosestMatch(normalized, allOptions);
  
  if (match) {
    // Find the canonical type for the matched synonym
    for (const [canonicalType, synonyms] of Object.entries(TYPE_MAPPINGS)) {
      if (synonyms.includes(match)) {
        return canonicalType as z.infer<typeof ClothingTypeEnum>;
      }
    }
    // If match is a canonical type
    if (CLOTHING_TYPE_VALUES.includes(match as z.infer<typeof ClothingTypeEnum>)) {
      return match as z.infer<typeof ClothingTypeEnum>;
    }
  }
  
  return 'other';
};

// Normalize subtype
export const normalizeSubType = (type: string | null | undefined, subType: string | null | undefined): string | null => {
  if (!subType) return null;
  
  const normalized = normalizeString(subType);
  if (!normalized) return null;
  
  // If the subtype is actually a type, return null
  if (CLOTHING_TYPE_VALUES.includes(normalized as z.infer<typeof ClothingTypeEnum>)) {
    return null;
  }
  
  // Check if the subtype is a synonym of the main type
  if (type && TYPE_MAPPINGS[type as keyof typeof TYPE_MAPPINGS]?.includes(normalized)) {
    return null;
  }
  
  return normalized.slice(0, MAX_SUBTYPE_LENGTH);
};

// Main validation function
export const validateClothingItem = (item: unknown): ClothingItem => {
  // First normalize the type and subtype
  const normalizedItem = {
    ...(item as Record<string, unknown>),
    type: normalizeClothingType((item as any)?.type),
    subType: normalizeSubType((item as any)?.type, (item as any)?.subType)
  };
  
  // Then validate against the schema
  return ClothingItemSchema.parse(normalizedItem);
};

export function validateClothingItems(items: Partial<ClothingItem>[]): ClothingItem[] {
  try {
    // First normalize all items
    const normalizedItems = items.map(item => ({
      ...item,
      type: normalizeClothingType(item.type),
      subType: normalizeSubType(item.type, item.subType)
    }));
    
    // Then validate against the schema
    return z.array(ClothingItemSchema).parse(normalizedItems);
  } catch (error) {
    if (error instanceof z.ZodError) {
      console.error('Validation error:', error.errors.map(e => `${e.path.join('.')}: ${e.message}`).join(', '));
    }
    throw error;
  }
}

export const validateOpenAIAnalysis = (analysis: unknown): OpenAIClothingAnalysis => {
  return OpenAIClothingAnalysisSchema.parse(analysis);
};

export function validateOutfit(outfit: unknown): Outfit {
  return OutfitSchema.parse(outfit);
}

export function validateUserProfile(profile: unknown): UserProfile {
  return UserProfileSchema.parse(profile);
}

// API Response helpers
export function createSuccessResponse<T>(data: { data: T; message?: string }): SuccessResponse<T> {
  return {
    success: true,
    data: data.data,
    message: data.message
  };
}

export function createErrorResponse(error: { code: string; message: string; details?: unknown }): ErrorResponse {
  return {
    success: false,
    data: null,
    error: `${error.code}: ${error.message}${error.details ? ` - ${JSON.stringify(error.details)}` : ''}`
  };
}

// Conversion functions
export function convertOpenAIAnalysisToClothingItem(
  analysis: OpenAIClothingAnalysis,
  userId: string,
  imageUrl: string
): Omit<ClothingItem, 'id'> {
  const basicMetadata = analysis.metadata?.basicMetadata ? {
    width: analysis.metadata.basicMetadata.width,
    height: analysis.metadata.basicMetadata.height,
    orientation: analysis.metadata.basicMetadata.orientation,
    dateTaken: analysis.metadata.basicMetadata.dateTaken,
    deviceModel: analysis.metadata.basicMetadata.deviceModel,
    gps: analysis.metadata.basicMetadata.gps,
    flashUsed: analysis.metadata.basicMetadata.flashUsed,
    imageHash: null
  } : undefined;

  const colorAnalysis: ColorAnalysis = {
    dominant: analysis.dominantColors,
    matching: analysis.matchingColors
  };

  return {
    name: analysis.type,
    type: normalizeClothingType(analysis.type),
    subType: normalizeSubType(analysis.type, analysis.subType),
    color: analysis.dominantColors[0]?.name || 'unknown',
    season: analysis.season,
    imageUrl,
    tags: [],
    style: analysis.style,
    userId,
    dominantColors: analysis.dominantColors,
    matchingColors: analysis.matchingColors,
    occasion: analysis.occasion,
    createdAt: Date.now(),
    updatedAt: Date.now(),
    brand: analysis.brand,
    metadata: {
      originalType: analysis.type,
      analysisTimestamp: Date.now(),
      basicMetadata,
      colorAnalysis,
      naturalDescription: null
    }
  };
}

// Validation error helpers
export function isValidationError(error: unknown): error is z.ZodError {
  return error instanceof z.ZodError;
}

export function formatValidationError(error: z.ZodError): { code: string; message: string; details: unknown } {
  return {
    code: 'VALIDATION_ERROR',
    message: 'Invalid data format',
    details: error.errors.map(e => ({
      path: e.path.join('.'),
      message: e.message
    }))
  };
}

// Type guard
export const isClothingItem = (item: unknown): item is ClothingItem => {
  try {
    validateClothingItem(item);
    return true;
  } catch {
    return false;
  }
}; 

File: backend/src/utils/validation.py
from typing import Optional, Dict, List, Any, Union
from unidecode import unidecode
import re
from difflib import SequenceMatcher
from ..custom_types.wardrobe import ClothingType, ClothingItem
import time
import json
import traceback
import logging
import uuid

# Set up logging
logger = logging.getLogger(__name__)

# Maximum lengths for strings
MAX_TYPE_LENGTH = 32
MAX_SUBTYPE_LENGTH = 64

# Canonical type mappings with synonyms
TYPE_MAPPINGS: Dict[str, List[str]] = {
    "shirt": [
        "t-shirt", "tee", "tshirt", "tee shirt", "v neck", "v-neck", "vneck", 
        "tank top", "polo", "blouse", "button down", "button-down", "button up",
        "button-up", "oxford", "dress shirt", "casual shirt", "formal shirt"
    ],
    "pants": [
        "jeans", "trousers", "slacks", "chinos", "khakis", "leggings", "joggers",
        "cargo pants", "dress pants", "formal pants", "casual pants", "shorts",
        "bermuda shorts", "capri pants", "culottes", "palazzo pants"
    ],
    "dress": [
        "gown", "frock", "sundress", "cocktail dress", "maxi", "mini",
        "evening dress", "formal dress", "casual dress", "party dress",
        "shift dress", "a-line dress", "bodycon dress", "wrap dress"
    ],
    "skirt": [
        "mini skirt", "maxi skirt", "pleated", "a-line", "pencil skirt",
        "midi skirt", "circle skirt", "wrap skirt", "denim skirt", "leather skirt"
    ],
    "jacket": [
        "denim", "blazer", "bomber", "parka", "windbreaker", "coat", "raincoat",
        "leather jacket", "suit jacket", "sports jacket", "cardigan", "sweater jacket",
        "puffer jacket", "down jacket", "fleece jacket", "vest", "waistcoat"
    ],
    "sweater": [
        "sweatshirt", "hoodie", "cardigan", "pullover", "jumper",
        "turtleneck", "mock neck", "crew neck", "v-neck sweater",
        "cable knit", "fisherman sweater", "chunky knit", "cashmere"
    ],
    "shoes": [
        "sneakers", "boots", "loafers", "sandals", "heels", "flats", "pumps",
        "oxfords", "derby", "mules", "espadrilles", "moccasins", "slip-ons",
        "ankle boots", "knee-high boots", "chelsea boots", "running shoes",
        "athletic shoes", "trainers", "stilettos", "wedges", "platforms"
    ],
    "accessory": [
        "bag", "purse", "scarf", "hat", "belt", "jewelry", "watch",
        "sunglasses", "gloves", "mittens", "socks", "tights", "stockings",
        "handbag", "backpack", "tote", "clutch", "necklace", "bracelet",
        "earrings", "ring", "tie", "bow tie", "pocket square", "cufflinks"
    ],
    "other": []
}

# Additional subtype mappings for common variations
SUBTYPE_MAPPINGS: Dict[str, List[str]] = {
    "shirt": [
        "v-neck", "crew neck", "turtle neck", "mock neck", "polo",
        "button-down", "button-up", "oxford", "dress", "casual"
    ],
    "pants": [
        "skinny", "straight", "bootcut", "wide leg", "flared",
        "cargo", "dress", "casual", "formal", "denim", "leather"
    ],
    "dress": [
        "maxi", "mini", "midi", "a-line", "bodycon", "shift",
        "wrap", "cocktail", "evening", "casual", "formal"
    ],
    "skirt": [
        "mini", "maxi", "midi", "a-line", "pencil", "pleated",
        "circle", "wrap", "denim", "leather"
    ],
    "jacket": [
        "denim", "leather", "bomber", "blazer", "parka", "windbreaker",
        "rain", "puffer", "down", "fleece", "vest"
    ],
    "sweater": [
        "v-neck", "crew neck", "turtle neck", "mock neck", "cardigan",
        "pullover", "hoodie", "cable knit", "chunky", "cashmere"
    ],
    "shoes": [
        "sneakers", "boots", "loafers", "sandals", "heels", "flats",
        "pumps", "oxfords", "derby", "mules", "espadrilles"
    ]
}

def normalize_string(input_str: Optional[str]) -> str:
    """Normalize a string by converting to lowercase, removing special characters, etc."""
    if not input_str:
        return ""
    
    # Convert to lowercase and normalize unicode
    normalized = unidecode(input_str.lower().strip())
    
    # Remove special characters but keep spaces and hyphens
    normalized = re.sub(r'[^a-z0-9\s-]', '', normalized)
    
    # Collapse multiple spaces and hyphens
    normalized = re.sub(r'[\s-]+', ' ', normalized)
    
    return normalized

def calculate_similarity(str1: str, str2: str) -> float:
    """Calculate similarity between two strings using SequenceMatcher."""
    return SequenceMatcher(None, str1, str2).ratio()

def find_closest_match(input_str: str, options: List[str], threshold: float = 0.8) -> Optional[str]:
    """Find the closest match from a list of options using string similarity."""
    if not input_str or not options:
        return None
    
    normalized_input = normalize_string(input_str)
    best_match = None
    best_score = threshold
    
    for option in options:
        normalized_option = normalize_string(option)
        score = calculate_similarity(normalized_input, normalized_option)
        
        if score > best_score:
            best_score = score
            best_match = option
    
    return best_match

def normalize_clothing_type(item_type: str, sub_type: Optional[str] = None) -> ClothingType:
    """
    Normalize clothing type with special handling for shorts and shoes.
    """
    if not item_type:
        logger.warning("Empty type string, defaulting to OTHER")
        return ClothingType.OTHER

    # Handle ClothingType enum format (e.g., "ClothingType.SHIRT" -> "shirt")
    if 'clothingtype.' in item_type.lower():
        enum_value = item_type.split('.')[-1].lower()
        # Map enum values to valid ClothingType enum values
        enum_mappings = {
            'shirt': ClothingType.SHIRT,
            'pants': ClothingType.PANTS, 
            'shoes': ClothingType.SHOES,
            'jacket': ClothingType.JACKET,
            'dress': ClothingType.DRESS,
            'sweater': ClothingType.SWEATER,
            'blouse': ClothingType.BLOUSE,
            'hoodie': ClothingType.HOODIE,
            'polo': ClothingType.POLO,
            'jeans': ClothingType.JEANS,
            'shorts': ClothingType.SHORTS,
            'skirt': ClothingType.SKIRT,
            'boots': ClothingType.BOOTS,
            'sneakers': ClothingType.SNEAKERS,
            'heels': ClothingType.HEELS,
            'blazer': ClothingType.BLAZER,
            'coat': ClothingType.COAT,
            'accessory': ClothingType.ACCESSORY,
            'other': ClothingType.OTHER
        }
        if enum_value in enum_mappings:
            return enum_mappings[enum_value]
        else:
            logger.warning(f"Unknown enum value {enum_value}, defaulting to OTHER")
            return ClothingType.OTHER

    # Convert to lowercase for case-insensitive matching
    type_lower = item_type.lower()
    sub_type_lower = sub_type.lower() if sub_type else None

    # Special case for shorts - check both type and subtype
    if (type_lower == "pants" and sub_type_lower == "shorts") or \
       (type_lower == "shorts") or \
       (sub_type_lower == "shorts"):
        return ClothingType.SHORTS

    # Special case for shoes - always return SHOES type regardless of subtype
    if type_lower == "shoes" or type_lower == "boots" or type_lower == "sneakers" or type_lower == "sandals":
        return ClothingType.SHOES

    # Map common variations to standard types with improved mapping
    type_mapping: Dict[str, ClothingType] = {
        # Shirts
        "shirt": ClothingType.SHIRT,
        "dress_shirt": ClothingType.DRESS_SHIRT,
        "t-shirt": ClothingType.T_SHIRT,
        "tshirt": ClothingType.T_SHIRT,
        "blouse": ClothingType.BLOUSE,
        "tank_top": ClothingType.TANK_TOP,
        "tank": ClothingType.TANK_TOP,
        "crop_top": ClothingType.CROP_TOP,
        "crop": ClothingType.CROP_TOP,
        "polo": ClothingType.POLO,
        "top": ClothingType.SHIRT,
        
        # Bottoms
        "pants": ClothingType.PANTS,
        "trousers": ClothingType.PANTS,
        "jeans": ClothingType.JEANS,
        "chinos": ClothingType.CHINOS,
        "slacks": ClothingType.SLACKS,
        "joggers": ClothingType.JOGGERS,
        "sweatpants": ClothingType.SWEATPANTS,
        "shorts": ClothingType.SHORTS,
        "skirt": ClothingType.SKIRT,
        "mini_skirt": ClothingType.MINI_SKIRT,
        "midi_skirt": ClothingType.MIDI_SKIRT,
        "maxi_skirt": ClothingType.MAXI_SKIRT,
        "pencil_skirt": ClothingType.PENCIL_SKIRT,
        
        # Dresses
        "dress": ClothingType.DRESS,
        "gown": ClothingType.DRESS,
        "sundress": ClothingType.SUNDRESS,
        "cocktail_dress": ClothingType.COCKTAIL_DRESS,
        "maxi_dress": ClothingType.MAXI_DRESS,
        "mini_dress": ClothingType.MINI_DRESS,
        
        # Outerwear
        "jacket": ClothingType.JACKET,
        "coat": ClothingType.COAT,
        "blazer": ClothingType.BLAZER,
        "vest": ClothingType.VEST,
        "sweater": ClothingType.SWEATER,
        "jumper": ClothingType.SWEATER,
        "hoodie": ClothingType.HOODIE,
        "cardigan": ClothingType.CARDIGAN,
        
        # Shoes
        "shoes": ClothingType.SHOES,
        "dress_shoes": ClothingType.DRESS_SHOES,
        "loafers": ClothingType.LOAFERS,
        "sneakers": ClothingType.SNEAKERS,
        "boots": ClothingType.BOOTS,
        "sandals": ClothingType.SANDALS,
        "heels": ClothingType.HEELS,
        "flats": ClothingType.FLATS,
        
        # Accessories
        "accessory": ClothingType.ACCESSORY,
        "jewelry": ClothingType.JEWELRY,
        "bag": ClothingType.BAG,
        "hat": ClothingType.HAT,
        "scarf": ClothingType.SCARF,
        "belt": ClothingType.BELT,
        "watch": ClothingType.WATCH,
        "other": ClothingType.OTHER
    }

    # Try to find a match in the mapping
    for key, value in type_mapping.items():
        if key in type_lower:
            return value

    # If no match found, try to match against enum values
    try:
        return ClothingType(type_lower)
    except ValueError:
        logger.warning(f"Could not normalize type '{item_type}', defaulting to OTHER")
        return ClothingType.OTHER

def normalize_subtype(sub_type: Optional[str]) -> Optional[str]:
    """
    Normalize clothing subtype.
    """
    if not sub_type:
        return None

    # Remove special characters and normalize
    normalized = ''.join(c for c in str(sub_type) if c.isalnum() or c.isspace() or c == '-')
    return normalized.strip()

def validate_clothing_item(item: Dict[str, Any]) -> ClothingItem:
    """
    Validate and normalize a clothing item.
    """
    # Normalize type and subtype
    item_type = normalize_clothing_type(((item.get('type', '') if item else '') if item else ''), item.get('subType'))
    sub_type = normalize_subtype((item.get('subType') if item else None))

    # Update the item with normalized values
    item['type'] = item_type
    item['subType'] = sub_type

    # Create and validate ClothingItem
    try:
        return ClothingItem(**item)
    except Exception as e:
        logger.error(f"Error validating clothing item: {str(e)}")
        raise ValueError(f"Invalid clothing item: {str(e)}")

def validate_outfit_requirements(outfit: ClothingItem) -> bool:
    """
    Validate that an outfit meets all requirements.
    """
    # Check for required pieces using core categories
    from .layering import get_core_category
    
    top_items = [item for item in outfit.items if get_core_category(item.type) == CoreCategory.TOP]
    bottom_items = [item for item in outfit.items if get_core_category(item.type) == CoreCategory.BOTTOM]
    dress_items = [item for item in outfit.items if get_core_category(item.type) == CoreCategory.DRESS]
    
    has_top = len(top_items) > 0
    has_bottom = len(bottom_items) > 0
    has_dress = len(dress_items) > 0
    has_mixed_bottoms = any(item.type == ClothingType.PANTS for item in outfit.items) and any(item.type == ClothingType.SHORTS for item in outfit.items)

    # Get style and season from metadata
    style = outfit.(metadata.get('style', '') if metadata else '').lower() if outfit.metadata else ''
    season = outfit.(metadata.get('season', '') if metadata else '').lower() if outfit.metadata else ''
    occasion = outfit.(metadata.get('occasion', '') if metadata else '').lower() if outfit.metadata else ''
    
    # Special handling for vacation outfits
    if occasion == 'vacation':
        # Vacation outfits can be more flexible
        if not (has_top or has_bottom):  # Allow single-piece outfits like dresses
            logger.warning("Vacation outfit should have at least one piece")
            return False
        
        # Allow mixing of casual pieces
        if has_mixed_bottoms:
            logger.info("Vacation outfit can mix pants and shorts")
            return True
        
        # Vacation-specific piece count
        piece_count = len(outfit.items)
        if piece_count < 1 or piece_count > 5:  # More flexible piece count for vacation
            logger.warning(f"Vacation outfit should have 1-5 pieces, got {piece_count}")
            return False
        
        return True

    # Regular outfit validation
    if has_dress:
        # If we have a dress, we don't need separate top and bottom
        if not has_dress:
            logger.warning("Outfit with dress should have at least one dress")
            return False
    else:
        # If no dress, we need both top and bottom
        if not has_top:
            logger.warning("Outfit missing a top (shirt, sweater, jacket, etc.)")
            return False

        if not has_bottom:
            logger.warning("Outfit missing pants, shorts, or skirt")
            return False

    if has_mixed_bottoms:
        logger.warning("Outfit cannot mix pants and shorts")
        return False

    # Define piece count ranges based on style and season
    piece_count = len(outfit.items)
    min_pieces = 2
    max_pieces = 6
    
    # Adjust for summer styles
    if season == 'summer' or 'summer' in style:
        min_pieces = 2
        max_pieces = 4
    
    # Adjust for winter styles
    if season == 'winter' or 'winter' in style:
        min_pieces = 3
        max_pieces = 6
    
    # Adjust for specific styles
    if 'minimalist' in style:
        min_pieces = 2
        max_pieces = 4
    elif 'layered' in style:
        min_pieces = 4
        max_pieces = 6
    
    # Check piece count
    if piece_count < min_pieces or piece_count > max_pieces:
        logger.warning(f"Outfit must have {min_pieces}-{max_pieces} pieces for {style} style in {season} season, got {piece_count}")
        return False

    return True

def is_clothing_item(item: Any) -> bool:
    """Type guard to check if an object is a valid clothing item."""
    try:
        validate_clothing_item(item)
        return True
    except Exception:
        return False 
